src/coreclr/nativeaot/Runtime/i386/MiscStubs.asm

```diff
@@ -37,6 +38,237 @@ ProbeLoop:
+    and     ecx, 63
+
+    cmp     ecx, 32
+    jae     LRshMORE32
+
+    ;; Handle shifts of between bits 0 and 31
+    shrd    eax, edx, cl
+    sar     edx, cl
+    ret
+
+LRshMORE32:
+    ;; Handle shifts of between bits 32 and 63
+    ;; The x86 shift instructions only use the lower 5 bits.
+    mov     eax, edx
+    sar     edx, 31
+    sar     eax, cl
+    ret
+RhpLRsh ENDP
+
+;; *********************************************************************/
+;;  LRsz:
+;; Purpose:
+;;    Does a unsigned Long Shift Right
+;;    Shifts a long right any number of bits.
+;;
+;;        NOTE:  This routine has been adapted from the Microsoft CRTs.
+;;
+;; Entry:
+;;    EDX:EAX - long value to be shifted
+;;        ECX - number of bits to shift by
+;;
+;; Exit:
+;;    EDX:EAX - shifted value
+;;
+RhpLRsz PROC public
+    ;; Reduce shift amount mod 64
+    and     ecx, 63
+
+    cmp     ecx, 32
+    jae     LRszMORE32
+
+    ;; Handle shifts of between bits 0 and 31
+    shrd    eax, edx, cl
+    shr     edx, cl
+    ret
+
+LRszMORE32:
+    ;; Handle shifts of between bits 32 and 63
+    ;; The x86 shift instructions only use the lower 5 bits.
+    mov     eax, edx
+    xor     edx, edx
+    shr     eax, cl
+    ret
+RhpLRsz ENDP
+
+;;
+;; Following helpers are unoptimized, forward to C library, and should
+;; eventually be revisited.
+;;
+
+EXTERN __dtol3 : PROC
+
+RhpDbl2Lng PROC public
+    movsd   xmm0, qword ptr [esp+4]
+    call    __dtol3
+    ret 8
+RhpDbl2Lng ENDP
+
+RhpDbl2Int PROC public
+    movsd   xmm0, qword ptr [esp+4]
+    cvttsd2si eax, xmm0
+    ret 8
+RhpDbl2Int ENDP
+
+RhpDbl2UInt PROC public
+    movsd   xmm0, qword ptr [esp+4]
+    call    __dtol3
+    ret 8
+RhpDbl2UInt ENDP
+
+;EXTERN __ltod3 : PROC
+
+RhpLng2Dbl PROC public
+    fild    qword ptr [esp+4]
+    ;mov     edx, dword ptr [esp+8]
+    ;mov     ecx, dword ptr [esp+4]
+    ;call    __ltod3
+    ;movsd   qword ptr [esp+4], xmm0
+    ;fld     qword ptr [esp+4]
+    ret     8
+RhpLng2Dbl ENDP
+
+EXTERN __ultod3 : PROC
+
+RhpULng2Dbl PROC public
+    mov     edx, dword ptr [esp+8]
+    mov     ecx, dword ptr [esp+4]
+    call    __ultod3
+    movsd   qword ptr [esp+4], xmm0
+    fld     qword ptr [esp+4]
+    ret     8
+RhpULng2Dbl ENDP
+
+EXTERN __alldiv : PROC
+
+RhpLDiv PROC public
+    push    dword ptr [esp+8]
+    push    dword ptr [esp+8]
+    push    dword ptr [esp+24]
+    push    dword ptr [esp+24]
+    call    __alldiv
+    ret     16
+RhpLDiv ENDP
+
+EXTERN __allrem : PROC
+
+RhpLMod PROC public
+    push    dword ptr [esp+8]
+    push    dword ptr [esp+8]
+    push    dword ptr [esp+24]
+    push    dword ptr [esp+24]
+    call    __allrem
+    ret     16
+RhpLMod ENDP
+
+EXTERN __aulldiv : PROC
+
+RhpULDiv PROC public
+    push    dword ptr [esp+8]
+    push    dword ptr [esp+8]
+    push    dword ptr [esp+24]
+    push    dword ptr [esp+24]
+    call    __aulldiv
+    ret     16
+RhpULDiv ENDP
+
+EXTERN __aullrem : PROC
+
+RhpULMod PROC public
```

Comment:
How does this work for divide by zero?

Reply:
The divide-by-zero check is on the managed side (`ULMod` in MathHelpers.cs).